backend/src/controller/donors.controllers.js

```diff
@@ -0,0 +1,156 @@
+const donorService = require("../services/donor.service.js");
+const Joi = require("joi");
+
+async function addDonor(req, res) {
+    // Joi validation
+    const schema = Joi.object().keys({
+        id: Joi.string().required(),
+        name: Joi.string()
+            .regex(/^[a-zA-Z]+$/)
+            .required()
+            .messages({
+                "string.pattern.base":
+                    "Name should not contain numbers or special characters",
+            }),
+        blood_group: Joi.string().required(),
+        address: Joi.string().required(),
+        contact: Joi.number().required(),
+        last_donated: Joi.string().required(),
+    });
+
+    if (schema.validate(req.body).error) {
+        console.log(schema.validate(req.body).error.details);
+        return res.status(422).send({ error: schema.validate(req.body).error.details });
+    }
+
+    try {
+        const name = await donorService.addDonor(req.body);
+        res.status(200).send(`Donor ${name} added successfully`);
+    } catch (err) {
+        res.status(401).send(err.message);
+    }
+}
+
+async function getDonors(req, res) {
+    if (req.params.searchTerm) {
+        // Joi validations
+        // for filtered search
+        const schema = Joi.object({
+            searchTerm:Joi.string.required(),       //required
+            page: Joi.number().optional(),          //optional input
+        });
+
+        if (schema.validate(req.query).error) {
+            console.log(schema.validate(req.body).error.details);
+            return res.status(422).send({ error: schema.validate(req.query).error.details });
+        }
+        const page = parseInt(req.query.page) || 1;
+        const query = req.params.searchTerm;
+        try {
+            const donors = await donorService.getFilteredDonors(page, query);
+            res.status(200).send(donors);
+        } catch (err) {
+            console.log(err.message);
+            res.status(401).send(err.message);
+        }
+    }
+    else{
+        // Joi validations
+        const schema = Joi.object({
+            page: Joi.number().optional(),
+        });
+    
+        if (schema.validate(req.query).error) {
+            console.log(schema.validate(req.query).error.details);
+            return res.status(422).send({ error: schema.validate(req.query).error.details });
+        }
+    
+        const page = parseInt(req.query.page) || 1;
+        try {
+            const donor = await donorService.getDonors(page);
+            res.status(200).send(donor);
+        } catch (err) {
+            console.log(err.message);
+            res.status(500).json({ error: err.message });
+        }
+    }
+}
+
+async function getFilteredDonors(req, res) {
```

Comment:
remove this one

Reply:
removed getFilteredDonors, now filteredDonorsData will be fetched from the single api "getDonors"